<launch>
    
    <!--- Sim Time -->
    <param name="/use_sim_time" value="true" />
    
    <arg name="file_path" default="$(find traversability_mapping)/rosbag"/>

    

    <arg name="lego" default="false"/>

    <arg name="full_info" default="/full_cloud_info"/>



        <!-- Rellis 3D -->

	<arg name="america" default="false"/>
	
	<arg name="path" default="/home/safe/Desktop/bags" />

	
	<!-- Ingeniarius -->
      <arg name="frame_id"       default="lslidar_link"/>
      <arg name="sensor_name"    default="front_lslidar"/> 

       <arg name="ingeniarius" default="false"/>

  

	<group if="$(arg lego)">


        <!--Image transports-->
    <node pkg="image_transport" 
		  type="republish" 
		  name="realsense_rgb" 
		  args="compressed in:=/realsense/color/image_raw
		  		raw out:=/realsense/color/image_raw"/> 
	
    <node pkg="image_transport" 
		  type="republish" 
		  name="realsense_depth" 
		  args="compressedDepth in:=/realsense/aligned_depth_to_color/image_raw
		        raw out:=/realsense/aligned_depth_to_color/image_raw"/>

	<node pkg="image_transport" 
	  type="republish" 
	  name="mynteye_left" 
	  args="compressed in:=mynteye/left_rect/image_rect
	  		raw out:=mynteye/left_rect/image_rect"/> 
	
	<node pkg="image_transport" 
	  type="republish" 
	  name="mynteye_right" 
	  args="compressed in:=mynteye/right_rect/image_rect
	  		raw out:=mynteye/right_rect/image_rect"/> 


	<!-- Generate Realsense's Point Cloud from Depth-->
	<node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager" />
  	<node pkg="nodelet" type="nodelet" name="realsense_depth2cloud"
  	      args="load depth_image_proc/point_cloud_xyzrgb nodelet_manager">
  	  <remap from="rgb/camera_info" to="/realsense/color/camera_info"/>
  	  <remap from="rgb/image_rect_color" to="/realsense/color/image_raw"/>

	  <remap from="depth_registered/image_rect" to="/realsense/aligned_depth_to_color/image_raw"/>
  	  <remap from="depth_registered/points" to="/realsense/depth/points"/>
  	</node>
	

    <node pkg="rosbag" type="play" name="choupal_0_bag" args=" --clock  /home/safe/Desktop/bags/localization_15Hz.bag /home/safe/Desktop/bags/choupal_0.bag /home/safe/Desktop/bags/choupal_1.bag /home/safe/Desktop/bags/choupal_2.bag "   		output="screen"/>

     


	
    <!--- Lidar Odometry -->    
    <node pkg="lego_loam" type="imageProjection"    name="imageProjection"    output="screen"/>
    <node pkg="lego_loam" type="featureAssociation" name="featureAssociation" output="screen"/>
    <node pkg="lego_loam" type="mapOptmization"     name="mapOptmization"     output="screen"/>
	
    <node pkg="lego_loam" type="transformFusion"    name="transformFusion"    output="screen"/>
	


   
    <node pkg="tf" type="static_transform_publisher" name="camera_init_to_map"  args="0 0 0 0  0 0 /map /camera_init 10" /> //1.57 0 1.57 //deu bem com 0 0 0
   

    <node pkg="tf" type="static_transform_publisher" name="base_link_to_camera" args="0.0404 0 0.264 0 0 0 /camera /livox_frame   10" /> // -1.57 -1.57 0 // -1,57095 0 0 camera base_link

    <node pkg="tf" type="static_transform_publisher" name="velodyne_base_link" args="0 0 0 0 0 0  /livox_frame /base_link2 10" /> //1.57 0 0 //1.57095 0 0 base_link - base_link2

    

    <!--- Run Rviz (launch-prefix="nice" for lower cpu usage)-->
    <node pkg="rviz" type="rviz" name="rviz" output="log" args="-d $(find traversability_mapping)/launch/include/traversability_mapping.rviz" />
   

 	<arg name="model" default="$(find traversability_mapping)/urdf/sensor_box.urdf"/>
	<param name="robot_description" textfile="$(arg model)" />
	<node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
            <remap from="joint_state" to="joint_state"/>
 	</node> 


	</group>


   <group if="$(arg america)">

      <node pkg="tf" type="static_transform_publisher" name="camera_init_to_map"  args="0 0 0 1.570795   0 1.570795 /map /camera_init 10" /> //1.57 0 1.57
   

      <node pkg="tf" type="static_transform_publisher" name="base_link_to_camera" args="0 0 0 -1.570795 -1.570795 0 /camera /base_link   10" /> // -1.57 -1.57 0

      <node pkg="tf" type="static_transform_publisher" name="velodyne_base_link" args="0 0 0 0 0 0  /base_link /base_link2 10" /> //1.57 0 0

	

    <!--<remap from="/os1_cloud_node/points" to="/livox/lidar"/> -->
    <remap from="/velodyne_points" to="/livox/lidar"/> 

    <node pkg="rosbag" type="play" name="choupal_0_bag" args="--clock  /home/safe/Desktop/bags/20200213_trail_2.bag"  output="screen"/>


    <node pkg="rviz" type="rviz" name="rviz" output="log" args="-d $(find traversability_mapping)/launch/include/traversability_mappingamerica.rviz" />

 <!--- Lidar Odometry -->    
    <node pkg="lego_loam" type="imageProjection"    name="imageProjection"    output="screen"/>
    <node pkg="lego_loam" type="featureAssociation" name="featureAssociation" output="screen"/>
    <node pkg="lego_loam" type="mapOptmization"     name="mapOptmization"     output="screen"/>
    <node pkg="lego_loam" type="transformFusion"    name="transformFusion"    output="screen"/>
	




	</group>


    <group if="$(arg ingeniarius)">

    <!-- <node pkg="rosbag" type="play" name="ingeniarius" args="  /home/safe/safeforest/src/safeforest_datasets/ros_bags/_2023-02-23-12-23-46_0.bag"  output="screen"/> -->

     <node pkg="rosbag" type="play" name="ingeniarius2022" args="--clock  /home/safe/gitsinced/src/ranger_datasets/dataset/2022_04_04_ranger_trial1.bag" output="screen" />

    <node pkg="tf" type="static_transform_publisher" name="camera_init_to_map"  args="0 0 0 0  0 0 /map /camera_init 10" /> //1.57 0 1.57 //deu bem com 0 0 0
   
    <node pkg="tf" type="static_transform_publisher" name="camera_base" args="0 0 0 0 0 0 /camera /base_link   10" /> // -1.57 -1.57 0 // -1,57095 0 0 camera base_link

    <node pkg="tf" type="static_transform_publisher" name="base_base" args="0 0 0 0 0 0 /base_link /bobcat_base   10" /> // -1.57 -1.57 0 // -1,57095 0 0 camera base_link

    <node pkg="tf" type="static_transform_publisher" name="base_lidar" args="0 0 0 0 0 0 /bobcat_base /front_lslidar   10" /> // -1.57 -1.57 0 // -1,57095 0 0 camera base_link	

    <node pkg="tf" type="static_transform_publisher" name="velodyne_base_link" args="0 0 0 0 0 0  /front_lslidar /base_link2 10" /> //1.57 0 0 //1.57095 0 0 base_link - base_link2

    <node pkg="rviz" type="rviz" name="rviz" output="log" args="-d $(find traversability_mapping)/launch/include/traversability_mappingamerica.rviz" />

    <!-- Front Ranger dynamic tf  -->
    <node pkg="ranger_datasets" type="republisher_odom_tf" name="robot_tool_republisher_odom_tf"> <!--output="screen"-->
        <param name="input_posEndEffector"   type="string"  value="endEffector_odom"/>  
        <param name="child_frame_id"         type="string"  value="end_effector_link"/>  
        <param name="frame_id"               type="string"  value="bobcat_base"/>  		
    </node>

    

   <!-- GET THE BACK LIDAR DATA POINTCLOUD  -->
    <node pkg="lslidar_c16_decoder" type="lslidar_c16_decoder_node" name="back_lslidar_c16_decoder_node"> <!--output="screen"-->
	    <param name="child_frame_id" value="back_lslidar"/>
	    <param name="point_num" value="2000"/>
	    <param name="channel_num" value="8"/>
	    <param name="angle_disable_min" value="0"/>
	    <param name="angle_disable_max" value="0"/>
	    <param name="min_range" value="0.15"/>
	    <param name="max_range" value="150.0"/>
	    <param name="frequency" value="10.0"/>
	    <param name="publish_point_cloud" value="true"/>
	    <param name="publish_channels" value="false"/>
	    <param name="use_gps_ts" value="false"/>

	    <!--output topic remapping for the greater scheme of things -->
	    <remap from="scan"           to="back_lslidar_scan"/>
	    <remap from="scan_channel"   to="back_lslidar_scan_channel"/>
	    <remap from="layer_num"      to="back_lslidar_layer_num"/>
	    <remap from="lslidar_packet" to="back_lslidar_packet"/>
	    <remap from="lslidar_sweep"  to="back_lslidar_sweep"/>
	    <remap from="lslidar_point_cloud" to="back_lslidar_point_cloud"/>
    </node>

    <!-- GET THE FRONT LIDAR DATA POINTCLOUD  -->
    <node pkg="lslidar_c16_decoder" type="lslidar_c16_decoder_node" name="front_lslidar_c16_decoder_node"> <!--output="screen"-->
	    <param name="child_frame_id" value="front_lslidar"/>
	    <param name="point_num" value="2000"/>
	    <param name="channel_num" value="8"/>
	    <param name="angle_disable_min" value="0"/>
	    <param name="angle_disable_max" value="0"/>
	    <param name="min_range" value="0.15"/>
	    <param name="max_range" value="150.0"/>
	    <param name="frequency" value="10.0"/>
	    <param name="publish_point_cloud" value="true"/>
	    <param name="publish_channels" value="false"/>
	    <param name="use_gps_ts" value="false"/>

	    <!--output topic remapping for the greater scheme of things -->
	    <remap from="scan"           to="front_lslidar_scan"/>
	    <remap from="scan_channel"   to="front_lslidar_scan_channel"/>
	    <remap from="layer_num"      to="front_lslidar_layer_num"/>
	    <remap from="lslidar_packet" to="front_lslidar_packet"/>  
	    <remap from="lslidar_sweep"  to="front_lslidar_sweep"/>
	    <remap from="lslidar_point_cloud" to="front_lslidar_point_cloud"/>	    
	</node>
    

    <remap from="/front_lslidar_point_cloud" to="/livox/lidar"/> 
    <remap from="/imu_rion" to="/imu/data"/>

    <node pkg="lego_loam" type="imageProjection"    name="imageProjection"    output="screen"/>
    <node pkg="lego_loam" type="featureAssociation" name="featureAssociation" output="screen"/>
    <node pkg="lego_loam" type="mapOptmization"     name="mapOptmization"     output="screen"/>
    <node pkg="lego_loam" type="transformFusion"    name="transformFusion"    output="screen"/>

    	<!-- base_to_back_bumper -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_to_back_bumper" 			args="-1.435 0 1.14 -3.14159265359 0 0 bobcat_base back_bumper_link"/>
	
	<!-- back_bumper_to_back_lslidar old config-->
	<node pkg="tf2_ros" type="static_transform_publisher" name="back_bumper_to_back_lslidar" 	args="0 0 0.215 0 0.25 0 back_bumper_link back_lslidar"/>


 	<!-- back_bumper_to_back_lslidar with NEW Lidar device-->
 <!-- 	<node pkg="tf2_ros" type="static_transform_publisher" name="back_bumper_to_back_lslidar" 	args="0 0 0.225 1.57 0 0.25 back_bumper_link back_lslidar"/> -->

	<!-- base_to_back_box -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_to_back_box" 	args="-1.40 0 1.12 -3.14159265359 0 0 bobcat_base back_box_link"/>

    <!-- base_to_sensor_box -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_to_sensor_box" 	args="0.4 0 2.2 0 0.48 0 bobcat_base sensor_box_link"/>

    <!-- base_to_sensing_kit -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_to_sensing_kit" 	args="0.5275 0.001 2.368 0 0 0 bobcat_base base_sensing_kit"/>

    <!-- base_sensing_kit_to_dalsa -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_sensing_kit_to_dalsa" 	args="0.137 0 -0.055 0 0.3 0 base_sensing_kit dalsa_link"/>

    <!-- base_sensing_kit_to_flir_ax8 -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_sensing_kit_to_flir_ax8" 	args="0.162 0 -0.12 0 0.3 0 base_sensing_kit flir_ax8_link"/>

    <!-- base_sensing_kit_to_realsense_USB -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_sensing_kit_to_realsense" 	args="0.153 0 -0.15 0 1.15 0 base_sensing_kit camera_link"/>

	<!-- back_bumper_to_back_center_framos -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="back_bumper_to_back_center_framos" 	args="0 0 0.05 0 0 0 back_bumper_link camera_back_center_link"/>

	<!-- back_bumper_to_back_right_framos -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="back_bumper_to_back_right_framos" 	args="-0.10 0.67 0.05 2.356194 0.66322 1.570796 back_bumper_link camera_back_right_link"/>

	<!-- back_bumper_to_back_left_framos -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="back_bumper_to_back_left_framos" 	args="-0.10 -0.67 0.05 -2.356194 0.66322 1.570796 back_bumper_link camera_back_left_link"/>
	
    <!-- framos_camera_tool_to_end_effector FRAMOS camera-->
	<!-- <node pkg="tf2_ros" type="static_transform_publisher" name="base_to_realsense_tool" 	args="0.545 0 1.125 0 0 0 end_effector_link camera_front_tool_link"/> -->

    <!-- base_sensing_kit_to_front_lslidar old-->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_sensing_kit_to_front_lslidar" 	args="0 0 0.051 0 0.52 0 base_sensing_kit front_lslidar"/>  
  

  	<!-- front_bumper_to_back_lslidar with NEW Lidar device-->
 	<!-- <node pkg="tf2_ros" type="static_transform_publisher" name="base_sensing_kit_to_front_lslidar" 	args="0 0 0.051 1.57 0 0.52 base_sensing_kit front_lslidar"/>  
 -->
    <!-- dalsa_to_dalsa_optical_frame -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="dalsa_to_dalsa_optical_frame" 	args="0 0 0 -1.57079 0 -1.57079 dalsa_link dalsa_optical_frame"/>       

    <!-- IMU UM7 -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_to_imu" 	args="0.15 0 2.2 0 -1.45640673352 0.78539816339 bobcat_base imu_link"/>      

    <!-- IMU RION -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_to_imu_rion" 	args="0 0 0 1.75 0 0 bobcat_base imu_rion_link"/>

	  <!-- IMU RION OLD-->
	<!-- <node pkg="tf2_ros" type="static_transform_publisher" name="base_to_imu_rion" 	args="-0.1975 0 1.755 0 0 0 bobcat_base imu_rion_link"/> -->

    <!-- GPS -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_to_gps" 	args="0.0 0 2.0 0 0 0 bobcat_base gps_link"/>

    <!-- Radar -->
	<node pkg="tf2_ros" type="static_transform_publisher" name="base_to_radar" 	args="-0.33 0 1.9 3.09 0 0 bobcat_base navtech"/>	


    </group>

    <!--- Run Traversability Mapping -->    
    <node pkg="traversability_mapping" type="traversability_filter"  name="traversability_filter"  output="screen">
	<param name="topico_info" value = "$(arg full_info)"/>
    </node>
    <node pkg="traversability_mapping" type="traversability_map"     name="traversability_map"     output="screen"/>
    <node pkg="traversability_mapping" type="traversability_prm"     name="traversability_prm"     output="screen"/>
    <!-- <node pkg="traversability_mapping" type="traversability_path"    name="traversability_path"    output="screen"/> -->

    <!--- Run Move Base -->
    <!--<include file="$(find traversability_mapping)/launch/include/move_base.launch" /> -->

</launch>
